src/keystore.cpp

```diff
@@ -89,6 +89,40 @@ bool CBasicKeyStore::AddSpendingKey(const libzcash::SpendingKey &sk)
     LOCK(cs_SpendingKeyStore);
     auto address = sk.address();
     mapSpendingKeys[address] = sk;
-    mapNoteDecryptors.insert(std::make_pair(address, ZCNoteDecryption(sk.viewing_key())));
+    mapNoteDecryptors.insert(std::make_pair(address, ZCNoteDecryption(sk.receiving_key())));
     return true;
 }
+
+bool CBasicKeyStore::AddViewingKey(const libzcash::ViewingKey &vk)
+{
+    LOCK(cs_SpendingKeyStore);
+    auto address = vk.address();
+    mapViewingKeys[address] = vk;
+    mapNoteDecryptors.insert(std::make_pair(address, ZCNoteDecryption(vk.sk_enc)));
+    return true;
+}
+
+bool CBasicKeyStore::RemoveViewingKey(const libzcash::ViewingKey &vk)
+{
+    LOCK(cs_SpendingKeyStore);
+    mapViewingKeys.erase(vk.address());
```

Comment:
doesn't the element of `mapNoteDecryptors` also need to be erased?

Reply:
No; see the comment about this in `src/gtest/test_keystore.cpp`. The only time we use this function is when we are adding the spending key for a viewing key, so the decryptor doesn't need to be uncached.

Reply:
Well perhaps at least add some comment about this in `RemoveViewingKey` declaration? It can be confusing that a method of this name only removes *part* of the viewing key.

Reply:
Also, isn't it slightly better when adding the spending key to readd the note decryptor? Just on the off case that it was wrong in the viewing key?

Reply:
I just checked the call chain again, and this is what happens:

- `CWallet::AddZKey(key)` is called, which calls:
  - `CCryptoKeyStore::AddSpendingKey(key)`, which:
    - saves the spending key
    - calls `mapNoteDecryptors.insert(std::make_pair(address, ZCNoteDecryption(vk.sk_enc)))` which will do nothing because [the key already exists](http://www.cplusplus.com/reference/map/map/insert/).
  - `return false` if the above fails for some reason.
  - `RemoveViewingKey(key.viewing_key())` if there is a viewing key to remove.

We definitely want to remove the viewing key after adding the spending key, because if the latter fails, we want to ensure we still have the viewing key. So we don't want to remove the cached note decryptor from memory when removing the viewing key, because then we won't have one at all.

> Also, isn't it slightly better when adding the spending key to readd the note decryptor? Just on the off case that it was wrong in the viewing key?

If it *did* happen to be wrong (which would require having used a malicious viewing key, since a purely corrupted one would cause the checksum to fail), the problem would be resolved on restart (when the note decryptors would be cached from the spending key). I don't see this as enough of a problem to alter the way the internal APIs are encapsulated.

Reply:
Please file a ticket about that problem, though.